presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java

```diff
@@ -266,6 +263,25 @@ default void finishDelete(ConnectorSession session, ConnectorTableHandle tableHa
     }
 
     /**
+     * Notifies when select query starts
+     *
+     * @return serializable {@link ConnectorTableHandle}. This handle will be brought back in {@link #finishSelect(ConnectorSession, ConnectorTableHandle)} if null returned, no finishSelect will be called.
+     */
+    default ConnectorTableHandle beginSelect(ConnectorSession session, ConnectorTableHandle tableHandle, Optional<ConnectorTableLayoutHandle> layoutHandle, Collection<ColumnHandle> columnHandles)
```

Comment:
I'm not swapping table handle, I'm just registering it to use it as an information keeper for `finishSelect`. It's here: https://github.com/prestodb/presto/pull/5122/files#diff-dfecb0b24d9d5c723d964cb59873b44eR606
And why I use `TableHandle` instead of custom structure: to not introduce new dependency for Connector SPI and to reuse existing Metadata`s methods for finding connector (`lookupConnectorFor` etc.).
The returned `TableHandle` is not passed to any other place but `finishSelect`.